crates/pkgcraft/src/eapi.rs

```diff
@@ -24,32 +25,79 @@ use crate::shell::hooks::HookBuilder;
 use crate::shell::operations::{Operation, OperationKind};
 use crate::shell::phase::Phase;
 
-peg::parser!(grammar parse() for str {
-    // EAPIs must not begin with a hyphen, dot, or plus sign.
-    pub(super) rule eapi() -> &'input str
-        = s:$(quiet!{
-            ['a'..='z' | 'A'..='Z' | '0'..='9' | '_']
-            ['a'..='z' | 'A'..='Z' | '0'..='9' | '+' | '_' | '.' | '-']*
-        } / expected!("EAPI"))
-        { s }
-
-    rule single_quotes<T>(expr: rule<T>) -> T
-        = "\'" v:expr() "\'" { v }
+mod parser {
+    use winnow::combinator::{alt, delimited, trace};
+    use winnow::prelude::*;
+    use winnow::token::{one_of, take_while};
+
+    const APOSTROPHE: char = '\'';
+    const QUOTATION_MARK: char = '"';
+
+    /// Parse an eapi, specified in PMS 3.1.9.
+    fn eapi_value<'i>(input: &mut &'i str) -> ModalResult<&'i str> {
+        trace(
+            "eapi_value",
+            (
+                one_of(('a'..='z', 'A'..='Z', '0'..='9', '_')),
+                take_while(0.., ('a'..='z', 'A'..='Z', '0'..='9', '_', '+', '.', '-')),
+            )
+                .take(),
+        )
+        .parse_next(input)
+    }
 
-    rule double_quotes<T>(expr: rule<T>) -> T
-        = "\"" v:expr() "\"" { v }
+    /// Parse an eapi, optionally quoted with either single or double quotes.
+    pub(super) fn eapi<'i>(input: &mut &'i str) -> ModalResult<&'i str> {
+        trace(
+            "eapi",
+            alt((
+                eapi_value,
+                delimited(QUOTATION_MARK, eapi_value, QUOTATION_MARK),
+                delimited(APOSTROPHE, eapi_value, APOSTROPHE),
+            )),
+        )
+        .parse_next(input)
+    }
 
-    rule optionally_quoted<T>(expr: rule<T>) -> T
-        = s:expr() { s }
-        / s:double_quotes(<expr()>) { s }
-        / s:single_quotes(<expr()>) { s }
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+
+        #[test]
+        fn value() {
+            assert_eq!(eapi_value.parse("a"), Ok("a"));
```

Comment:
All these test lines should probably be collapsed using a loop for valid and another for invalid similar to what's done elsewhere.